src/compile.ts

```diff
@@ -817,10 +832,82 @@ export function compile(
         } else if (ts.isNonNullExpression(node)) {
           return toExpr(node.expression, scope);
         } else if (node.kind === ts.SyntaxKind.ThisKeyword) {
-          // assuming that this is used in a valid location, create a closure around that instance.
-          return ref(ts.factory.createIdentifier("this"));
+          return newExpr("ThisExpr", [
+            ts.factory.createArrowFunction(
+              undefined,
+              undefined,
+              [],
+              undefined,
+              undefined,
+              ts.factory.createIdentifier("this")
+            ),
+          ]);
+        } else if (
+          ts.isToken(node) &&
+          node.kind === ts.SyntaxKind.SuperKeyword
+        ) {
+          return newExpr("SuperKeyword", []);
         } else if (ts.isAwaitExpression(node)) {
           return newExpr("AwaitExpr", [toExpr(node.expression, scope)]);
+        } else if (ts.isClassDeclaration(node) || ts.isClassExpression(node)) {
+          return newExpr(
+            `Class${ts.isClassDeclaration(node) ? "Decl" : "Expr"}`,
+            [
+              // name
+              node.name ?? ts.factory.createIdentifier("undefined"),
+              // extends
+              node.heritageClauses?.flatMap((clause) =>
+                clause.token === ts.SyntaxKind.ExtendsKeyword &&
+                clause.types[0].expression !== undefined
+                  ? [toExpr(clause.types[0].expression, scope)]
+                  : []
+              )[0] ?? ts.factory.createIdentifier("undefined"),
```

Comment:
why `[0]`?

Reply:
In the TS AST, the `heritageClauses` can be `extends` or `implements` - an array. Here, I am using `flatMap` to filter out only the `extends` clause which will result in an array of length 1 or an empty array if there is no `extends` clause. `[0]` grabs the value if it exists, otherwise I returned `undefined`.